### ISSUE TITLE
[java] WhileLoopWithLiteralBoolean can be modified to support only while loops.

### ISSUE THREAD

**Issue by Fernal73** [pmd/pmd#2154]
**Affects PMD Version:** 
6.20.0

**Rule:**
WhileLoopWithLiteralBoolean

**Description:**
Add a property to suggest converting a do while loop to a while loop making the condition visible at the outset. This will be an optional feature permitting developers to exercise their preference for while loops as against do while.
The rule can be renamed, if needed.

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*

**Comment by adangel** [comment]
I think, this should be a rule on its own, since [WhileLoopWithLiteralBoolean](https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#whileloopwithliteralboolean) is very limited. It only looks a loops where the condition is a boolean literal constant.

Let's create an example: This new rule should detect following snippet:

```java
boolean someCondition = ...;
boolean someOtherCondition = ...;

do {
    // more code
} while (someCondition && someOtherCondition);
```

Hm... I'm not sure, what we would suggest here? Do-while loop execute the body at least once regardless of the condition. So just moving the condition to the outset will change the program...
The example from above with a while loop would look like this:

```java
boolean someCondition = ...;
boolean someOtherCondition = ...;

// more code - the duplicated loop body
// usually you would extract the loop body into a method a call the method here

while (someCondition && someOtherCondition) {
    // more code
};
```

Did I understand you correct? I find it much easier to work with examples than with textual descriptions of what the rule should do/not do...

**Comment by Fernal73** [comment]
What do you need an example for? A while loop simply has the condition set to true at the outset if you wish to execute it at least once---which is exactly the case with the while true as well. If you expect it to run only once every time, you don't need a loop. Period. That's pretty much evident from your explanation for the existing rule. The condition is irrelevant (in that case).

**Comment by oowekyala** [comment]
> What do you need an example for?
> I find it much easier to work with examples than with textual descriptions of what the rule should do/not do...

...

Besides, examples are needed to write tests, so at some point someone has to write them.

Anyway if I understand correctly, you want a new rule eg "DoLoopCanBeWhileLoop". AFAIK, `do` loops can always be replaced with `while` loops by duplicating the body before the loop. So I'm not sure how to formulate a rule that doesn't just flag all `do` loops. Your providing code examples would help.

Then, I don't understand the relation to WhileLoopWithLiteralBoolean, which has nothing to do with `do` loops. Also the title of the issue does not match the description, or the comments you posted.

>  If you expect it to run only once every time, you don't need a loop. Period. That's pretty much evident from your explanation for the existing rule. 

This seems more suited for a distinct rule, eg "LoopStatementDoesNotLoop".

**Comment by Fernal73** [comment]
This rule will replace all do while loops with a while loop. As I said , it's for programmers who prefer while loops with the condition upfront and shun do loops. It would be controversial and thus I preferred it as an optional choice in the existing rule.
